// @swimlane/ngx-graph/esm2015/lib/graph/layouts/colaForceDirected.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { id } from '../../utils/id';
import { d3adaptor } from 'webcola';
import * as d3Dispatch from 'd3-dispatch';
import * as d3Force from 'd3-force';
import * as d3Timer from 'd3-timer';
import { Subject } from 'rxjs';
/**
 * @record
 */
export function ColaForceDirectedSettings() { }
if (false) {
    /** @type {?|undefined} */
    ColaForceDirectedSettings.prototype.force;
    /** @type {?|undefined} */
    ColaForceDirectedSettings.prototype.forceModifierFn;
    /** @type {?|undefined} */
    ColaForceDirectedSettings.prototype.onTickListener;
    /** @type {?|undefined} */
    ColaForceDirectedSettings.prototype.viewDimensions;
}
/**
 * @record
 */
export function ColaGraph() { }
if (false) {
    /** @type {?} */
    ColaGraph.prototype.groups;
    /** @type {?} */
    ColaGraph.prototype.nodes;
    /** @type {?} */
    ColaGraph.prototype.links;
}
/**
 * @param {?} nodes
 * @param {?} nodeRef
 * @return {?}
 */
export function toNode(nodes, nodeRef) {
    if (typeof nodeRef === 'number') {
        return nodes[nodeRef];
    }
    return nodeRef;
}
export class ColaForceDirectedLayout {
    constructor() {
        this.defaultSettings = {
            force: d3adaptor(Object.assign({}, d3Dispatch, d3Force, d3Timer))
                .linkDistance(150)
                .avoidOverlaps(true),
            viewDimensions: {
                width: 600,
                height: 600,
                xOffset: 0
            }
        };
        this.settings = {};
        this.outputGraph$ = new Subject();
    }
    /**
     * @param {?} graph
     * @return {?}
     */
    run(graph) {
        this.inputGraph = graph;
        if (!this.inputGraph.clusters) {
            this.inputGraph.clusters = [];
        }
        this.internalGraph = {
            nodes: (/** @type {?} */ ([
                ...this.inputGraph.nodes.map(n => (Object.assign({}, n, { width: n.dimension ? n.dimension.width : 20, height: n.dimension ? n.dimension.height : 20 })))
            ])),
            groups: [
                ...this.inputGraph.clusters.map((cluster) => ({
                    padding: 5,
                    groups: cluster.childNodeIds
                        .map(nodeId => (/** @type {?} */ (this.inputGraph.clusters.findIndex(node => node.id === nodeId))))
                        .filter(x => x >= 0),
                    leaves: cluster.childNodeIds
                        .map(nodeId => (/** @type {?} */ (this.inputGraph.nodes.findIndex(node => node.id === nodeId))))
                        .filter(x => x >= 0)
                }))
            ],
            links: (/** @type {?} */ ([
                ...this.inputGraph.edges
                    .map(e => {
                    /** @type {?} */
                    const sourceNodeIndex = this.inputGraph.nodes.findIndex(node => e.source === node.id);
                    /** @type {?} */
                    const targetNodeIndex = this.inputGraph.nodes.findIndex(node => e.target === node.id);
                    if (sourceNodeIndex === -1 || targetNodeIndex === -1) {
                        return undefined;
                    }
                    return Object.assign({}, e, { source: sourceNodeIndex, target: targetNodeIndex });
                })
                    .filter(x => !!x)
            ])),
            groupLinks: [
                ...this.inputGraph.edges
                    .map(e => {
                    /** @type {?} */
                    const sourceNodeIndex = this.inputGraph.nodes.findIndex(node => e.source === node.id);
                    /** @type {?} */
                    const targetNodeIndex = this.inputGraph.nodes.findIndex(node => e.target === node.id);
                    if (sourceNodeIndex >= 0 && targetNodeIndex >= 0) {
                        return undefined;
                    }
                    return e;
                })
                    .filter(x => !!x)
            ]
        };
        this.outputGraph = {
            nodes: [],
            clusters: [],
            edges: [],
            edgeLabels: []
        };
        this.outputGraph$.next(this.outputGraph);
        this.settings = Object.assign({}, this.defaultSettings, this.settings);
        if (this.settings.force) {
            this.settings.force = this.settings.force
                .nodes(this.internalGraph.nodes)
                .groups(this.internalGraph.groups)
                .links(this.internalGraph.links)
                .alpha(0.5)
                .on('tick', () => {
                if (this.settings.onTickListener) {
                    this.settings.onTickListener(this.internalGraph);
                }
                this.outputGraph$.next(this.internalGraphToOutputGraph(this.internalGraph));
            });
            if (this.settings.viewDimensions) {
                this.settings.force = this.settings.force.size([
                    this.settings.viewDimensions.width,
                    this.settings.viewDimensions.height
                ]);
            }
            if (this.settings.forceModifierFn) {
                this.settings.force = this.settings.forceModifierFn(this.settings.force);
            }
            this.settings.force.start();
        }
        return this.outputGraph$.asObservable();
    }
    /**
     * @param {?} graph
     * @param {?} edge
     * @return {?}
     */
    updateEdge(graph, edge) {
        /** @type {?} */
        const settings = Object.assign({}, this.defaultSettings, this.settings);
        if (settings.force) {
            settings.force.start();
        }
        return this.outputGraph$.asObservable();
    }
    /**
     * @param {?} internalGraph
     * @return {?}
     */
    internalGraphToOutputGraph(internalGraph) {
        this.outputGraph.nodes = internalGraph.nodes.map(node => (Object.assign({}, node, { id: node.id || id(), position: {
                x: node.x,
                y: node.y
            }, dimension: {
                width: (node.dimension && node.dimension.width) || 20,
                height: (node.dimension && node.dimension.height) || 20
            }, transform: `translate(${node.x - ((node.dimension && node.dimension.width) || 20) / 2 || 0}, ${node.y -
                ((node.dimension && node.dimension.height) || 20) / 2 || 0})` })));
        this.outputGraph.edges = internalGraph.links
            .map(edge => {
            /** @type {?} */
            const source = toNode(internalGraph.nodes, edge.source);
            /** @type {?} */
            const target = toNode(internalGraph.nodes, edge.target);
            return Object.assign({}, edge, { source: source.id, target: target.id, points: [
                    ((/** @type {?} */ (source.bounds))).rayIntersection(target.bounds.cx(), target.bounds.cy()),
                    ((/** @type {?} */ (target.bounds))).rayIntersection(source.bounds.cx(), source.bounds.cy())
                ] });
        })
            .concat(internalGraph.groupLinks.map(groupLink => {
            /** @type {?} */
            const sourceNode = internalGraph.nodes.find(foundNode => ((/** @type {?} */ (foundNode))).id === groupLink.source);
            /** @type {?} */
            const targetNode = internalGraph.nodes.find(foundNode => ((/** @type {?} */ (foundNode))).id === groupLink.target);
            /** @type {?} */
            const source = sourceNode || internalGraph.groups.find(foundGroup => ((/** @type {?} */ (foundGroup))).id === groupLink.source);
            /** @type {?} */
            const target = targetNode || internalGraph.groups.find(foundGroup => ((/** @type {?} */ (foundGroup))).id === groupLink.target);
            return Object.assign({}, groupLink, { source: source.id, target: target.id, points: [
                    ((/** @type {?} */ (source.bounds))).rayIntersection(target.bounds.cx(), target.bounds.cy()),
                    ((/** @type {?} */ (target.bounds))).rayIntersection(source.bounds.cx(), source.bounds.cy())
                ] });
        }));
        this.outputGraph.clusters = internalGraph.groups.map((group, index) => {
            /** @type {?} */
            const inputGroup = this.inputGraph.clusters[index];
            return Object.assign({}, inputGroup, { dimension: {
                    width: group.bounds ? group.bounds.width() : 20,
                    height: group.bounds ? group.bounds.height() : 20
                }, position: {
                    x: group.bounds ? group.bounds.x + group.bounds.width() / 2 : 0,
                    y: group.bounds ? group.bounds.y + group.bounds.height() / 2 : 0
                } });
        });
        this.outputGraph.edgeLabels = this.outputGraph.edges;
        return this.outputGraph;
    }
    /**
     * @param {?} draggingNode
     * @param {?} $event
     * @return {?}
     */
    onDragStart(draggingNode, $event) {
        /** @type {?} */
        const nodeIndex = this.outputGraph.nodes.findIndex(foundNode => foundNode.id === draggingNode.id);
        /** @type {?} */
        const node = this.internalGraph.nodes[nodeIndex];
        if (!node) {
            return;
        }
        this.draggingStart = { x: node.x - $event.x, y: node.y - $event.y };
        node.fixed = 1;
        this.settings.force.start();
    }
    /**
     * @param {?} draggingNode
     * @param {?} $event
     * @return {?}
     */
    onDrag(draggingNode, $event) {
        if (!draggingNode) {
            return;
        }
        /** @type {?} */
        const nodeIndex = this.outputGraph.nodes.findIndex(foundNode => foundNode.id === draggingNode.id);
        /** @type {?} */
        const node = this.internalGraph.nodes[nodeIndex];
        if (!node) {
            return;
        }
        node.x = this.draggingStart.x + $event.x;
        node.y = this.draggingStart.y + $event.y;
    }
    /**
     * @param {?} draggingNode
     * @param {?} $event
     * @return {?}
     */
    onDragEnd(draggingNode, $event) {
        if (!draggingNode) {
            return;
        }
        /** @type {?} */
        const nodeIndex = this.outputGraph.nodes.findIndex(foundNode => foundNode.id === draggingNode.id);
        /** @type {?} */
        const node = this.internalGraph.nodes[nodeIndex];
        if (!node) {
            return;
        }
        node.fixed = 0;
    }
}
if (false) {
    /** @type {?} */
    ColaForceDirectedLayout.prototype.defaultSettings;
    /** @type {?} */
    ColaForceDirectedLayout.prototype.settings;
    /** @type {?} */
    ColaForceDirectedLayout.prototype.inputGraph;
    /** @type {?} */
    ColaForceDirectedLayout.prototype.outputGraph;
    /** @type {?} */
    ColaForceDirectedLayout.prototype.internalGraph;
    /** @type {?} */
    ColaForceDirectedLayout.prototype.outputGraph$;
    /** @type {?} */
    ColaForceDirectedLayout.prototype.draggingStart;
}
